### PR TITLE
fix: Nullable arrays not being parsed as `null`

### PR DESCRIPTION
**Pull request by msfstef**
Addresses https://github.com/electric-sql/electric/issues/1597

I also made the types include the possibility of `null` being a value that needs to be parsed and changed the parsers accordingly.

I don't love that currently if a field is not nullable and `null` is received it tries to force it to a field (like `null` to `0` if it's a number), and this logic definitely does not work on arrays marked as `NOT NULL` that are `null`. Are we against throwing for those cases?

### CODE REVIEW

**Comment by kevin-dp** [comment]
> if a field is not nullable and null is received

This feels like it would be a bug in Electric.
Is this possible? How can this happen?

**Comment by msfstef** [comment]
@kevin-dp it should not be possible, which is why I'm wondering if we should be throwing instead of forcing it to a non-null value

**Comment by kevin-dp** [comment]
> @kevin-dp it should not be possible, which is why I'm wondering if we should be throwing instead of forcing it to a non-null value

Yep, i'd much rather throw an error saying we got an unexpected null value from Electric for a non-nullable column rather than forcing it into the right type which is just very likely to lead to problems or unexpected behavior somewhere else.

**Review by kevin-dp — CHANGES_REQUESTED** [review]
I've added some comments and requested some changes because i don't think we need to handle `null` values everywhere as is done currently. We should be able to limit handling of `null` values to only the `makeNullableParser` function.

File: packages/typescript-client/src/parser.ts [hunk 1]
```diff
@@ -1,7 +1,10 @@
 import { ColumnInfo, Message, Schema, Value } from './types'
 
-export type ParseFunction = (
-  value: string,
+type NullToken = null | `NULL`
+type Token = Exclude<string, NullToken>
+type NullableToken = Token | NullToken
+export type ParseFunction<T extends Token | NullableToken = Token> = (
```
Comment (kevin-dp) [inline packages/typescript-client/src/parser.ts:6]: `Token | NullableToken` is redundant since `Token` is already part of `NullableToken`, so should just be `NullableToken`.

Comment (kevin-dp) [reply packages/typescript-client/src/parser.ts:6]: I'm not a fan of using generics for representing two versions of this function type, it complicates the understanding a bit.
imo it's easier to understand if we define them as two separate types:
```ts
export type ParseFunction = (value: Token, additionalInfo?: Omit<ColumnInfo, `type` | `dims`>) => Value
export type NullableParseFunction = (value: NullableToken, additionalInfo?: Omit<ColumnInfo, `type` | `dims`>) => Value
```

**Review by kevin-dp — COMMENTED** [review]
Good work!
I like how you improved it now but we should add a unit test that checks that it now correctly parses `null` values on nullable array columns and perhaps a second test that checks that the parser complains if you try to parse a null value on a non-nullable column.

**Review by kevin-dp — APPROVED** [review]
